controllers/spiaccesstokenbinding_controller.go

```diff
@@ -298,26 +299,33 @@ func (r *SPIAccessTokenBindingReconciler) linkToken(ctx context.Context, sp serv
 			r.updateBindingStatusError(ctx, binding, api.SPIAccessTokenBindingErrorReasonLinkedToken, err)
 			return nil, fmt.Errorf("failed to create the token: %w", err)
 		}
+		newTokenCreated = true
 	}
 
 	// we need to have this label so that updates to the linked SPIAccessToken are reflected here, too... We're setting
 	// up the watch to use the label to limit the scope...
 	if err := r.persistWithMatchingLabels(ctx, binding, token); err != nil {
+		// linking newly created token failed, lets cleanup it
+		if newTokenCreated {
+			log.FromContext(ctx).Error(err, "linking of the created token failed, cleaning up token %s/%s, error: %w", token.GetNamespace(), token.GetName(), err)
+			r.cleanupUnlinkedToken(ctx, token)
+		}
 		return nil, err
 	}
 
 	return token, nil
 }
 
 func (r *SPIAccessTokenBindingReconciler) persistWithMatchingLabels(ctx context.Context, binding *api.SPIAccessTokenBinding, token *api.SPIAccessToken) error {
+	originalBinding := binding.DeepCopy()
```

Comment:
If we modify the object to become incompatible, updating it with error status will fail as well. So it's better to try that with the original one, which probably is still valid.

Reply:
The update may fail because of 2 reasons - the object is deleted from the cluster in the meantime or is updated in the meantime (which is detected by comparing the `metadata.resourceVersion` with the true state in the cluster) or it doesn't conform to the schema. In this method we only update the labels which cannot cause the update failure.

Reply:
Not only labels, but fields as well - and that was exactly our case, setting `OAuthURL` to empty which is against the schema, so neither this update and error status updates worked. (with that fix we would have Bining to fall into error state and doesn't reconsiling in cycle)

Reply:
but does it make any difference if you use `originalBinding` or `binding` for the `updateStatusBindingError` call? IMHO, it doesn't make a difference...

Reply:
If it was wrong from the beginning, it should been fail earlier, probably...
Well, for sure we can't predict everything, but imo having better chances to "emergency landing" is worth it.